File: src/main/go/g0001_0100/s0098_validate_binary_search_tree/solution.go
package s0098_validate_binary_search_tree

// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search #Tree #Binary_Tree
// #Binary_Search_Tree #Data_Structure_I_Day_14_Tree #Level_1_Day_8_Binary_Search_Tree
// #Udemy_Tree_Stack_Queue #Top_Interview_150_Binary_Search_Tree #Big_O_Time_O(N)_Space_O(log(N))
// #2025_05_06_Time_0_ms_(100.00%)_Space_7.21_MB_(46.50%)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	return checkBST(root, nil, nil)
}

func checkBST(root *TreeNode, min, max *int) bool {
	if root == nil {
		return true
	}
	if min != nil && *min >= root.Val {
		return false
	}
	if max != nil && root.Val >= *max {
		return false
	}
	return checkBST(root.Left, min, &root.Val) &&
		checkBST(root.Right, &root.Val, max)

}

File: src/main/erlang/g0001_0100/s0098_validate_binary_search_tree/Solution.erl
% #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search #Tree #Binary_Tree
% #Binary_Search_Tree #Data_Structure_I_Day_14_Tree #Level_1_Day_8_Binary_Search_Tree
% #Udemy_Tree_Stack_Queue #Top_Interview_150_Binary_Search_Tree #Big_O_Time_O(N)_Space_O(log(N))
% #2025_01_14_Time_0_ms_(100.00%)_Space_73.36_MB_(100.00%)

%% Definition for a binary tree node.
%%
%% -record(tree_node, {val = 0 :: integer(),
%%                     left = null  :: 'null' | #tree_node{},
%%                     right = null :: 'null' | #tree_node{}}).

-spec is_valid_bst(Root :: #tree_node{} | null) -> boolean().
is_valid_bst(Tree) ->
   is_valid_bst(null, Tree,  null).

in_range(null, Middle, null) ->
    true;

in_range(Min, Middle, null) ->
    Min < Middle;

in_range(null, Middle, Max) ->
    Middle < Max;

in_range(Min, Middle, Max) ->
    (Min < Middle) and (Middle < Max).

is_valid_bst(Min, null, Max) ->
    true;
is_valid_bst(Min, #tree_node{val = Val, left = Left, right = Right}, Max) ->
    in_range(Min, Val, Max) andalso
            is_valid_bst(Min, Left, Val) andalso 
                is_valid_bst(Val, Right, Max).
